File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Common/FoxgloveJsonMessageParser.swift
//
//  FoxgloveJsonMessageParser.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Foxglove JSON メッセージのパーサ（サーバー → クライアント）
struct FoxgloveJsonMessageParser {
    /// JSON 文字列をパースしてメッセージ種別に応じた結果を返す
    /// - Parameter jsonString: 受信した JSON 文字列
    /// - Returns: パース結果。パースに失敗した場合は nil
    static func parseServerMessage(_ jsonString: String) -> FoxgloveServerJsonMessage? {
        guard let jsonData = jsonString.data(using: .utf8) else {
            return nil
        }
        return parseServerMessage(jsonData)
    }

    /// JSON データをパースしてメッセージ種別に応じた結果を返す
    /// - Parameter jsonData: 受信した JSON データ
    /// - Returns: パース結果。パースに失敗した場合は nil
    private static func parseServerMessage(_ jsonData: Data) -> FoxgloveServerJsonMessage? {
        guard let header = try? JSONDecoder().decode(FoxgloveMessageHeader.self, from: jsonData) else {
            return nil
        }

        let decoder = JSONDecoder()
        switch header.op {
        case .serverInfo:
            guard let info = try? decoder.decode(FoxgloveServerInfo.self, from: jsonData) else { return nil }
            return .serverInfo(info)
        case .advertise:
            guard let advertise = try? decoder.decode(FoxgloveAdvertise.self, from: jsonData) else { return nil }
            return .advertise(advertise)
        case .unadvertise:
            guard let body = try? decoder.decode(FoxgloveUnadvertiseBody.self, from: jsonData) else { return nil }
            return .unadvertise(channelIds: body.channelIds)
        case .advertiseServices:
            guard let services = try? decoder.decode(FoxgloveAdvertiseServices.self, from: jsonData) else { return nil }
            return .advertiseServices(services)
        case .unadvertiseServices:
            guard let body = try? decoder.decode(FoxgloveUnadvertiseServicesBody.self, from: jsonData) else { return nil }
            return .unadvertiseServices(serviceIds: body.serviceIds)
        case .serviceCallFailure:
            guard let body = try? decoder.decode(FoxgloveServiceCallFailureBody.self, from: jsonData) else { return nil }
            return .serviceCallFailure(serviceId: body.serviceId, callId: body.callId, message: body.message)
        default:
            return .other
        }
    }
}

/// サーバーから受信した JSON メッセージのパース結果
enum FoxgloveServerJsonMessage {
    case serverInfo(FoxgloveServerInfo)
    case advertise(FoxgloveAdvertise)
    case unadvertise(channelIds: [UInt32])
    case advertiseServices(FoxgloveAdvertiseServices)
    case unadvertiseServices(serviceIds: [UInt32])
    case serviceCallFailure(serviceId: UInt32, callId: UInt32, message: String)
    case other
}

// MARK: - Private body types for partial parsing

private struct FoxgloveUnadvertiseBody: Codable {
    let channelIds: [UInt32]
}

private struct FoxgloveUnadvertiseServicesBody: Codable {
    let serviceIds: [UInt32]
}

private struct FoxgloveServiceCallFailureBody: Codable {
    let serviceId: UInt32
    let callId: UInt32
    let message: String
}

// MARK: - Private

/// JSON メッセージの op フィールドのみをパースして種別を判別するための基底型
private struct FoxgloveMessageHeader: Codable {
    let op: FoxgloveMessageOperation
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_android_client/data/infrastructure/foxglove/common/FoxgloveJsonMessageParser.kt
package com.shusuke.raspberry_pi_android_client.data.infrastructure.foxglove.common

import com.shusuke.raspberry_pi_android_client.data.infrastructure.foxglove.server.FoxgloveAdvertise
import com.shusuke.raspberry_pi_android_client.data.infrastructure.foxglove.server.FoxgloveAdvertiseServices
import com.shusuke.raspberry_pi_android_client.data.infrastructure.foxglove.server.FoxgloveServerInfo
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Foxglove JSON メッセージのパーサ（サーバー → クライアント）
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md
 */
object FoxgloveJsonMessageParser {

    private val json = Json {
        ignoreUnknownKeys = true
    }

    /**
     * JSON 文字列をパースしてメッセージ種別に応じた結果を返す
     *
     * @param jsonString 受信した JSON 文字列
     * @return パース結果。パースに失敗した場合は null
     */
    fun parseServerMessage(jsonString: String): FoxgloveServerJsonMessage? =
        runCatching {
            val header = json.decodeFromString<FoxgloveMessageHeader>(jsonString)
            val op = FoxgloveMessageOperation.fromWireString(header.op)
                ?: return@runCatching FoxgloveServerJsonMessage.Other
            when (op) {
                FoxgloveMessageOperation.ServerInfo ->
                    FoxgloveServerJsonMessage.ServerInfo(
                        json.decodeFromString<FoxgloveServerInfo>(jsonString),
                    )
                FoxgloveMessageOperation.Advertise ->
                    FoxgloveServerJsonMessage.Advertise(
                        json.decodeFromString<FoxgloveAdvertise>(jsonString),
                    )
                FoxgloveMessageOperation.Unadvertise ->
                    FoxgloveServerJsonMessage.Unadvertise(
                        json.decodeFromString<FoxgloveUnadvertiseBody>(jsonString).channelIds,
                    )
                FoxgloveMessageOperation.AdvertiseServices ->
                    FoxgloveServerJsonMessage.AdvertiseServices(
                        json.decodeFromString<FoxgloveAdvertiseServices>(jsonString),
                    )
                FoxgloveMessageOperation.UnadvertiseServices ->
                    FoxgloveServerJsonMessage.UnadvertiseServices(
                        json.decodeFromString<FoxgloveUnadvertiseServicesBody>(jsonString).serviceIds,
                    )
                FoxgloveMessageOperation.ServiceCallFailure ->
                    json.decodeFromString<FoxgloveServiceCallFailureBody>(jsonString).let { body ->
                        FoxgloveServerJsonMessage.ServiceCallFailure(
                            serviceId = body.serviceId,
                            callId = body.callId,
                            message = body.message,
                        )
                    }
                FoxgloveMessageOperation.Subscribe,
                FoxgloveMessageOperation.Unsubscribe,
                -> FoxgloveServerJsonMessage.Other
            }
        }.getOrNull()

    val jsonInstance: Json get() = json
}

/**
 * サーバーから受信した JSON メッセージのパース結果
 */
sealed class FoxgloveServerJsonMessage {
    data class ServerInfo(val info: FoxgloveServerInfo) : FoxgloveServerJsonMessage()
    data class Advertise(val advertise: FoxgloveAdvertise) : FoxgloveServerJsonMessage()
    data class Unadvertise(val channelIds: List<UInt>) : FoxgloveServerJsonMessage()
    data class AdvertiseServices(val services: FoxgloveAdvertiseServices) : FoxgloveServerJsonMessage()
    data class UnadvertiseServices(val serviceIds: List<UInt>) : FoxgloveServerJsonMessage()
    data class ServiceCallFailure(
        val serviceId: UInt,
        val callId: UInt,
        val message: String,
    ) : FoxgloveServerJsonMessage()
    data object Other : FoxgloveServerJsonMessage()
}

@Serializable
private data class FoxgloveMessageHeader(val op: String)

@Serializable
private data class FoxgloveUnadvertiseBody(
    @SerialName("channelIds")
    val channelIds: List<UInt>,
)

@Serializable
private data class FoxgloveUnadvertiseServicesBody(
    @SerialName("serviceIds")
    val serviceIds: List<UInt>,
)

@Serializable
private data class FoxgloveServiceCallFailureBody(
    @SerialName("serviceId")
    val serviceId: UInt,
    @SerialName("callId")
    val callId: UInt,
    val message: String,
)
